src/vmm/src/signal_handler.rs

```diff
@@ -1,168 +1,254 @@
+        let si_code = info.ssi_code;
+
+        // For SIGSYS, we have some special logging.
+        if si_signo == SIGSYS {
+            log_sigsys_err(info);
+        }
+
+        // Increment the right metrics.
+        match si_signo {
+            SIGXFSZ => METRICS.signals.sigxfsz.inc(),
+            SIGXCPU => METRICS.signals.sigxcpu.inc(),
+            SIGBUS => METRICS.signals.sigbus.inc(),
+            SIGSEGV => METRICS.signals.sigsegv.inc(),
+            SIGPIPE => METRICS.signals.sigpipe.inc(),
+            SIGSYS => METRICS.seccomp.num_faults.inc(),
+            SIGHUP => METRICS.signals.sighup.inc(),
+            SIGILL => METRICS.signals.sigill.inc(),
+            other => {
+                // This should never be reached since we'll only receive the signals in the signalfd mask.
+                error!(
+                    "Received unexpected signal: {} which is not in signalfd mask.",
+                    other
+                );
+                exit_unexpected();
+            }
+        };
+
+        error!(
+            "Shutting down VM after intercepting signal {}, code {}.",
+            si_signo, si_code
+        );
+        // Write the metrics before exiting.
+        if let Err(e) = METRICS.write() {
+            error!("Failed to write metrics while stopping: {}", e);
+        }
+
+        // Safe because we're terminating the process anyway. We don't actually do anything when
+        // running unit tests.
+        #[cfg(not(test))]
+        unsafe {
+            _exit(i32::from(match si_signo {
+                SIGXFSZ => super::FC_EXIT_CODE_SIGXFSZ,
+                SIGXCPU => super::FC_EXIT_CODE_SIGXCPU,
+                SIGBUS => super::FC_EXIT_CODE_SIGBUS,
+                SIGSEGV => super::FC_EXIT_CODE_SIGSEGV,
+                SIGPIPE => super::FC_EXIT_CODE_SIGPIPE,
+                SIGSYS => super::FC_EXIT_CODE_BAD_SYSCALL,
+                SIGHUP => super::FC_EXIT_CODE_SIGHUP,
+                SIGILL => super::FC_EXIT_CODE_SIGILL,
+                _ => super::FC_EXIT_CODE_UNEXPECTED_ERROR,
+            }))
+        };
+    }
+}
+
+impl Subscriber for SignalManager {
+    fn process(&mut self, event: &EpollEvent, _evmgr: &mut EventManager) {
+        let event_set = event.event_set();
+        let source_fd = event.fd();
+
+        if event_set != EventSet::IN || source_fd != self.signal_fd.as_raw_fd() {
+            error!("Spurious EventManager event for handler: SignalManager");
+            return;
+        }
+
+        let siginfo_size = mem::size_of::<signalfd_siginfo>();
+        let mut read_buf: Vec<u8> = vec![0; siginfo_size];
+
+        // Even though we currently kill the process during signal handling, we should in theory
+        // read from the signalfd all the signals for future-proofing.
+        loop {
+            match self.signal_fd.read(&mut read_buf[..]) {
+                Ok(size) if size == siginfo_size => {
+                    match SignalInfoWrapper::from_slice(&read_buf) {
```

Comment:
I don't think this is worth the added complexity introduced by `SignalInfoWrapper`. Personally I would just read. It should be safe.

Reply:
As we discussed offline, there is not much added complexity and the `ByteValued` approach has some clear advantages:

 - does not copy the buffer, it uses the same memory as the byte slice, as opposed to the `ptr::read`
 - has some extra safety checks: checks that the data is correctly aligned and checks the length of the byte slice.